### ISSUE TITLE
Export HID++ library

### ISSUE THREAD

**Issue by hadess** [libratbag/libratbag#45]
libratbag currently has its own HID++ handling code, which supports both the 1.0 and 2.0 versions.

We have similar code in UPower to query battery status and solar charging support:
http://cgit.freedesktop.org/upower/tree/src/linux/hidpp-device.h

And we would want to integrate unifying receiver support in GNOME directly, to allow pairing and unpairing devices:
https://bugzilla.gnome.org/show_bug.cgi?id=756574

To avoid adding external dependencies, the new library could be a cut'n'paste library, or a git submodule. The API needed would be:
- enumerate
- get sysfs path (to match with udev, eventually), whether device is a touchpad, keyboard, mouse, etc.
- start pairing
- unpair a device
- get battery percentage, whether charging or not
- when supported, the luminosity hitting the solar charger

**Comment by bentiss** [comment]
There is actually two slightly different problems here:

- the receiver handling for enumerating devices, pairing and unpairing
- accessing devices capabilities and dealing with them

Both functionalities are relying on the HID++ protocol from Logitech. But I am not sure we want to provide a common API for both.
The problem I see is that if you enumerate devices from the receiver, you can not go from the returned device to an actual hidpp_device with the current code. And I am not sure you want to do that given the 2 use cases.

After brainstorming with @whot yesterday, we think will provide 2 sub-libraries, one for accessing the receiver, and one for accessing the functionalities you need for both hid++ 1.0 and 2.0 devices.

How does that sound?

**Comment by whot** [comment]
https://github.com/whot/libratbag/tree/wip/receiver-library
is the first attempt, let me know what you think.

This is a library for accessing Logitech Unified Receivers, and provides the basic functionality you asked for (minus the sysfs bits). The caller opens a hidraw node, constructs a receiver device, then does enumerate/open/close/disconnect on the device. The devices give you name, pid/vid and type.

Only a header file and a basic tool to show how it would work, no code to back it up yet.
If that's what you want, I can start working on that though.

**Comment by hadess** [comment]
That interface looks good to me, I'd just be missing a couple of things:
- something to filter unifying receivers with (a .rules files with a particular tag, or a function that tells me whether the usb device with this vid/pid is a unifying receiver). I can probably piggy-back on UPower's tags for that (UPOWER_BATTERY_TYPE=unifying)
- something to tell me whether a particular device is connected (it's at the receiver level, right?)
- a way to match connected devices in the library to connected devices in udev (if the device serial were exposed in both the kernel and the library, or the index were exposed in the kernel). This is to solve the edge case of 2 devices with the same name.

Thanks!

**Comment by whot** [comment]
- added lur_is_receiver(int vid, int pid)
- there's an activity counter we could expose, otherwise the receiver knows which device is connected, even if the device isn't active right now. that's what the enumerate() will give you, but if you need something specific let me know so we can be sure it's doable.
- the serial is easy enough to expose from liblur. I wonder if that's something a udev callout could set on the udev device? @bentiss, do yo have a better suggestion re:kernel maybe?

**Comment by bentiss** [comment]
> the serial is easy enough to expose from liblur. I wonder if that's something a udev callout could set on the udev device? @bentiss, do yo have a better suggestion re:kernel maybe?

Yes, it's easy to export the serial when we have the handle on the receiver.

The easiest way to know if a udev device is connected to a particular receiver when you have its udev handle is simply to look at the udev tree: the receiver attaches the input devices that are connected to itself below it's own device. For example when I have only one device attached to the receiver:

```
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/0003:046D:4041.005D
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/0003:046D:4041.005D/hidraw/hidraw2
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/0003:046D:4041.005D/input/input133
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/0003:046D:4041.005D/input/input133/event11
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/0003:046D:4041.005D/input/input133/mouse3
P: /devices/.../1-1:1.2/0003:046D:C52B.005C/hidraw/hidraw1
```

The receiver is the hidraw1 node, and the device is the hidraw2/event11 node (with the wireless PID 0x4041).

**Comment by whot** [comment]
https://github.com/whot/libratbag/tree/wip/receiver-library

updated the branch with an actual implementation. the lur-command tool works, at least in my testing here. lur-command list /dev/hidraw2 will list all devices paired with that receiver (even if they're turned off). Is that enough or do you want some activity check too?

HID++ has an activity counter but it'd require polling the device regularly to see if the counter increases. Alternatively, we could probably ping the device with some other request and check if it responds - if not, it's paired but not turned on or suspended.

**Comment by hadess** [comment]
@whot I would just need the serials exported in udev/sysfs, that way, I can find whether the device is active on the local machine or not (which is the same as connected/disconnected for 99% of cases), which would be good enough for me.

**Comment by hadess** [comment]
The serial fetching doesn't seem to work (both are connected, I'm using them):
```
$ lur-command list /dev/hidraw0
0: T650 (touchpad) serial 0
1: K750 (keyboard) serial 0
```

This fixes it:
```
diff --git a/src/liblur.c b/src/liblur.c
index cbe1c64..65fd0dd 100644
--- a/src/liblur.c
+++ b/src/liblur.c
@@ -228,6 +228,7 @@ lur_receiver_enumerate(struct lur_receiver *lur,
                        dev->vid = VENDOR_ID_LOGITECH;
                        dev->pid = wpid;
                        dev->type = type;
+                       dev->serial = serial;
                        dev->hidppidx = i;
                        dev->present = true;
                        list_insert(&lur->devices, &dev->node);
```

And the lur-command is really unforgiving:
```
$ lur-command --help
lur-command: lur-command.c:106: main: Assertion `argc > 2' failed.
Aborted (core dumped)
```

**Comment by whot** [comment]
squashed in, thanks. And yes, it's just running aborts right now for anything it can't handle, was a lot easier than proper error handling. But looks like we generally agree on this half of the library, so I can start adding those bits to it.

**Comment by whot** [comment]
alright, updated, with better handling now. afaict what is missing now is the kernel exporting the serial (@bentiss) and having this as a copy/paste library. atm it's a .so.

**Comment by whot** [comment]
Merged in pull #53
